server/src/server.rs

```diff
@@ -0,0 +1,709 @@
+					});
+
+					response.map(|()| hyper::Body::empty())
+				}
+				Err(e) => {
+					tracing::error!("Could not upgrade connection: {}", e);
+					hyper::Response::new(hyper::Body::from(format!("Could not upgrade connection: {}", e)))
+				}
+			};
+
+			async { Ok(response) }.boxed()
+		} else {
+			// The request wasn't an upgrade request; let's treat it as a standard HTTP request:
+			let data = http::HandleRequest {
+				methods: self.inner.methods.clone(),
+				resources: self.inner.resources.clone(),
+				max_request_body_size: self.inner.max_request_body_size,
+				max_response_body_size: self.inner.max_response_body_size,
+				max_log_length: self.inner.max_log_length,
+				batch_requests_supported: self.inner.batch_requests_supported,
+				logger: self.inner.logger.clone(),
+				conn: self.inner.conn.clone(),
+			};
+
+			Box::pin(http::handle_request(request, data).map(Ok))
+		}
+	}
+}
+
+/// This is a glorified select listening for new messages, while also checking the `stop_receiver` signal.
+struct Monitored<'a, F> {
+	future: F,
+	stop_monitor: &'a StopHandle,
+}
+
+impl<'a, F> Monitored<'a, F> {
+	fn new(future: F, stop_monitor: &'a StopHandle) -> Self {
+		Monitored { future, stop_monitor }
+	}
+}
+
+enum MonitoredError<E> {
+	Shutdown,
+	Selector(E),
+}
+
+struct Incoming(TcpListener);
+
+impl<'a> Future for Monitored<'a, Incoming> {
+	type Output = Result<(TcpStream, SocketAddr), MonitoredError<std::io::Error>>;
+
+	fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+		let this = Pin::into_inner(self);
+
+		if this.stop_monitor.shutdown_requested() {
+			return Poll::Ready(Err(MonitoredError::Shutdown));
+		}
+
+		this.future.0.poll_accept(cx).map_err(MonitoredError::Selector)
+	}
+}
+
+impl<'a, 'f, F, T, E> Future for Monitored<'a, Pin<&'f mut F>>
+where
+	F: Future<Output = Result<T, E>>,
+{
+	type Output = Result<T, MonitoredError<E>>;
+
+	fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+		let this = Pin::into_inner(self);
+
+		if this.stop_monitor.shutdown_requested() {
+			return Poll::Ready(Err(MonitoredError::Shutdown));
+		}
+
+		this.future.poll_unpin(cx).map_err(MonitoredError::Selector)
+	}
+}
+
+// Attempts to accept a new connection
+async fn try_accept_connection<S, Bd>(
+	socket: TcpStream,
+	service: S,
+	mut stop_handle: StopHandle,
+	curr_conns: usize,
+	max_conns: usize,
+) where
+	S: Service<hyper::Request<hyper::Body>, Response = hyper::Response<Bd>> + Send + 'static,
+	S::Error: Into<Box<dyn std::error::Error + Send + Sync>>,
+	S::Future: Send,
+	Bd: HttpBody + Send + 'static,
+	<Bd as HttpBody>::Error: Send + Sync + StdError,
+	<Bd as HttpBody>::Data: Send,
+{
+	let conn = hyper::server::conn::Http::new().serve_connection(socket, service).with_upgrades();
+
+	tokio::pin!(conn);
+
+	tokio::select! {
+		res = &mut conn => {
+			match res {
+				Ok(_) => tracing::info!("Accepting new connection {}/{}", curr_conns, max_conns),
```

Comment:
shall we log the `peer addr` here as well? I reckon that it might be useful

Reply:
That sounds like a good idea, we can always come with another PR 
Wondering if we should incorporate the `peer addr` or `connection address` to uniquely identify which logs are coming from which connections?
It might help us group errors and warnings that have a common root cause

Reply:
yeah, it was one thing missing from the logs when I tried to debug these `Transport(i/o error: unexpected end of file)` errors that we have seen plenty of

Reply:
let's do it in another PR